src/lib.rs

```diff
@@ -2,269 +2,225 @@
+                                archive_entry_gid(entry).try_into()?,
+                            )),
+                        )?;
+                        archive_read_data_skip(ar);
+                    } else {
+                        let file = std::fs::File::create(&file_path)?;
+                        nix::unistd::chown(
+                            &file_path,
+                            Some(nix::unistd::Uid::from_raw(
+                                archive_entry_uid(entry).try_into()?,
+                            )),
+                            Some(nix::unistd::Gid::from_raw(
+                                archive_entry_gid(entry).try_into()?,
+                            )),
+                        )?;
+                        nix::sys::stat::fchmod(
+                            file.as_raw_fd(),
+                            nix::sys::stat::Mode::from_bits(archive_entry_perm(entry)).unwrap(),
+                        )?;
+                        read_data_block(file, ar)?;
+                    }
+                }
+                ARCHIVE_EOF => {
+                    return Ok(());
+                }
+                _ => return Err(compress::Error::from(archive_error_string(ar))),
+            }
         }
+    })
+}
 
-        env::set_var(
-            "PATH",
-            format!(
-                "{}{}",
-                mocks_dir.display(),
-                &env::var("PATH")
-                    .map(|s| format!(":{}", s))
-                    .unwrap_or_default()
-            ),
-        );
-
-        Ok((mocks, calls))
-    }
-
-    fn assert_calls(p: &Path, expected: &[&str]) {
-        let mut content = String::default();
-        fs::File::open(p)
-            .unwrap()
-            .read_to_string(&mut content)
-            .unwrap();
-        for call in expected {
-            assert!(
-                content.contains(call),
-                format!(
-                    "uncompress did not call the expected: '{}'\nFull content:\n{}",
-                    call, content
-                )
-            );
+/// Reads the input buffer and extract a file to another buffer.
+///
+/// # Arguments
+///
+/// * `source` - Reader.
+/// * `target` - Writer.
+/// * `path` - File path to extract to the target.
+///
+/// # Example
+///
+/// uncompress_archive_file(std::io::Read, std::io::Write, path: &str)?;
+pub fn uncompress_archive_file<R: 'static, W>(
+    source: R,
+    target: W,
+    path: &str,
+) -> compress::Result<()>
+where
+    R: std::io::Read,
+    W: std::io::Write,
+{
+    run_with_archive(false, source, |ar, mut entry| unsafe {
+        loop {
+            if archive_read_next_header(ar, &mut entry) == ARCHIVE_OK {
+                let file_name = CStr::from_ptr(archive_entry_pathname(entry)).to_str()?;
+                if file_name == path {
+                    break;
+                }
+            } else {
+                return Err(compress::Error::from(archive_error_string(ar)));
+            }
```

Comment:
Looks like we can also use `handle_error` here and we also would lower one indentation level.